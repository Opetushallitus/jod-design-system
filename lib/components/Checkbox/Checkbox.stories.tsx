import { action } from '@storybook/addon-actions';
import { useArgs, useState } from '@storybook/preview-api';
import type { StoryObj } from '@storybook/react';
import type { TitledMeta } from '../../utils';
import { Checkbox, CheckboxProps } from './Checkbox';

const meta = {
  title: 'Forms/Checkbox',
  component: Checkbox,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof Checkbox>;

export default meta;

type Story = StoryObj<typeof meta>;
const backgrounds = {
  default: 'jod-bg-white',
  values: [
    { name: 'jod-bg-white', value: '#fff' },
    { name: 'jod-bg-1', value: '#f0f2f5' },
    { name: 'jod-bg-2', value: '#e4e6ec' },
  ],
};

const design = {
  type: 'figma',
  url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542-7171',
};
const label = 'Tuloksen sopivuus';
const render = (args: CheckboxProps) => {
  const [, setArgs] = useArgs();
  const [value, setValue] = useState<boolean>(args.checked);

  const onChange = (value: boolean) => {
    action('onChange')(value);
    setArgs({ checked: value });
    setValue(value);
  };

  return <Checkbox {...args} checked={value} onChange={() => onChange(!value)} />;
};

export const Available: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story: 'This is a available checkbox.',
      },
    },
  },
  args: {
    name: 'name',
    disabled: false,
    value: 'value',
    checked: false,
    label,
    ariaLabel: label,
  },
  render,
};

export const Selected: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story: 'This is a selected checkbox.',
      },
    },
  },
  args: {
    name: 'name',
    disabled: false,
    value: 'value',
    checked: true,
    label,
    ariaLabel: label,
  },
  render,
};

export const Bordered: Story = {
  parameters: {
    backgrounds,
    design,
    docs: {
      description: {
        story: 'This is a checkbox with bordered variant.',
      },
    },
  },
  args: {
    name: 'name',
    value: 'value',
    checked: false,
    label,
    ariaLabel: label,
    variant: 'bordered',
  },
  render,
};
export const BorderedAndDisabled: Story = {
  parameters: {
    backgrounds,
    design,
    docs: {
      description: {
        story: 'This is a disabled checkbox with bordered variant.',
      },
    },
  },
  args: {
    name: 'name',
    disabled: true,
    value: 'value',
    checked: false,
    label,
    ariaLabel: label,
    variant: 'bordered',
  },
  render,
};

export const Disabled: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story: 'This is a disabled checkbox.',
      },
    },
  },
  args: {
    name: 'name',
    disabled: true,
    value: 'value',
    checked: false,
    label,
    ariaLabel: label,
  },
  render,
};

export const WithNoVisibleLabel: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story: 'This is a checkbox which has no visible label (having aria-label).',
      },
    },
  },
  args: {
    name: 'name',
    value: 'value',
    checked: false,
    variant: 'bordered',
    ariaLabel: label,
  },
  render,
};

export const DisabledAndBorderedWithNoVisibleLabel: Story = {
  parameters: {
    backgrounds,
    design,
    docs: {
      description: {
        story: 'This is a disabled checkbox with bordered variant, which has no visible label (having aria-label).',
      },
    },
  },
  args: {
    name: 'name',
    disabled: true,
    value: 'value',
    checked: false,
    ariaLabel: label,
    variant: 'bordered',
  },
  render,
};

export const ComponentAsLabel: Story = {
  parameters: {
    backgrounds,
    design,
    docs: {
      description: {
        story: 'This is a disabled checkbox with bordered variant, which has no visible label (having aria-label).',
      },
    },
  },
  args: {
    name: 'name',
    value: 'value',
    checked: false,
    label: (
      <div className="ds:flex ds:flex-row ds:ml-2 ds:gap-3">
        <span className="ds:text-heading-2 ds:ml-3">{label}</span>
        <span className="ds:size-3 ds:rounded ds:bg-success ds:block ds:self-center" />
      </div>
    ),
    ariaLabel: label,
    variant: 'bordered',
  },
  render,
};

export const IndeterminateCheckbox: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story:
          'This is a indeterminate checkbox that cycles between unchecked → indeterminate → checked → unchecked states.',
      },
    },
  },
  args: {
    name: 'indeterminate-state-checkbox',
    value: 'indeterminate-state',
    checked: false,
    label: 'Indeterminate checkbox',
    ariaLabel: 'Indeterminate checkbox',
  },
  render: () => {
    const [state, setState] = useState<'unchecked' | 'indeterminate' | 'checked'>('unchecked');

    const cycleState = () => {
      if (state === 'unchecked') {
        setState('indeterminate');
      } else if (state === 'indeterminate') {
        setState('checked');
      } else {
        setState('unchecked');
      }
    };

    return (
      <div className="ds:flex ds:flex-col ds:gap-4">
        <Checkbox
          name="indeterminate-state-checkbox"
          value="indeterminate-state"
          checked={state === 'checked'}
          indeterminate={state === 'indeterminate'}
          onChange={cycleState}
          label="Indeterminate-state checkbox"
          ariaLabel="Indeterminate-state checkbox"
        />
        <Checkbox
          name="indeterminate-state-checkbox"
          value="indeterminate-state"
          checked={state === 'checked'}
          indeterminate={state === 'indeterminate'}
          onChange={cycleState}
          label="Indeterminate-state checkbox with Bordered"
          ariaLabel="Indeterminate-state checkbox"
          variant="bordered"
        />
      </div>
    );
  },
};

export const IndeterminateCheckboxWithChildren: Story = {
  parameters: {
    design,
    docs: {
      description: {
        story:
          'A practical example of the indeterminate state with a parent checkbox controlling multiple child checkboxes.',
      },
    },
  },
  args: {
    name: 'parent',
    value: 'parent',
    checked: false,
    label: 'Select all items',
    ariaLabel: 'Select all items',
  },
  render: () => {
    // Track states of children checkboxes
    const [childrenState, setChildrenState] = useState([false, false, false]);

    // Calculate parent state based on children
    const allChecked = childrenState.every((state) => state);
    const someChecked = childrenState.some((state) => state);
    const parentChecked = allChecked;
    const parentIndeterminate = someChecked && !allChecked;

    const handleParentChange = () => {
      const newState = !allChecked;
      setChildrenState(childrenState.map(() => newState));
    };

    const handleChildChange = (index: number) => {
      const newChildrenState = [...childrenState];
      newChildrenState[index] = !newChildrenState[index];
      setChildrenState(newChildrenState);
    };

    return (
      <div className="ds:flex ds:flex-col ds:gap-3">
        <Checkbox
          name="parent"
          value="parent"
          checked={parentChecked}
          indeterminate={parentIndeterminate}
          onChange={handleParentChange}
          label="Select all items"
          ariaLabel="Select all items"
        />
        <div className="ds:ml-6 ds:flex ds:flex-col ds:gap-2">
          {['Item 1', 'Item 2', 'Item 3'].map((item, index) => (
            <Checkbox
              key={item}
              name={`child-${index}`}
              value={`child-${index}`}
              checked={childrenState[index]}
              onChange={() => handleChildChange(index)}
              label={item}
              ariaLabel={item}
            />
          ))}
        </div>

        <div className="mt-10" />
        <Checkbox
          name="parent"
          value="parent"
          checked={parentChecked}
          indeterminate={parentIndeterminate}
          onChange={handleParentChange}
          label="Select all items (with Bordered)"
          ariaLabel="Select all items (with Bordered)"
          variant="bordered"
        />
        <div className="ds:ml-6 ds:flex ds:flex-col ds:gap-2">
          {['Item 1', 'Item 2', 'Item 3'].map((item, index) => (
            <Checkbox
              key={item}
              name={`child-${index}`}
              value={`child-${index}`}
              checked={childrenState[index]}
              onChange={() => handleChildChange(index)}
              label={item}
              ariaLabel={item}
              variant="bordered"
            />
          ))}
        </div>
      </div>
    );
  },
};
