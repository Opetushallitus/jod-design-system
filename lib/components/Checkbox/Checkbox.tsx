import React from 'react';
import { cx } from '../../cva';

export interface CheckboxProps {
  /** Name for the component */
  name: string;
  /** Disabled state for the component */
  disabled?: boolean;
  /** Value for the component */
  value: string;
  /** Checked state for the component */
  checked: boolean;
  /** Change event for the component */
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /** Text for the component */
  label?: React.ReactNode;
  /** Aria label for the component */
  ariaLabel: string;
  /** Additional class name for the component */
  className?: string;
  /** Component variant */
  variant?: 'default' | 'bordered';
}

/**
 * Checkboxes allow users to select multiple items from a list of individual items, or to mark one individual item as selected.
 */
export const Checkbox = ({
  name,
  disabled,
  value,
  checked,
  onChange,
  label,
  ariaLabel,
  className,
  variant = 'default',
}: CheckboxProps) => {
  const id = React.useId();
  const isLabelValidElement = React.isValidElement(label);

  const borderVariantClassnames = {
    'ds-border ds-border-accent': variant === 'bordered',
    'ds-border ds-border-inactive-gray': variant === 'bordered' && disabled,
    'ds-border-0': variant === 'default',
  };

  return (
    <div className={cx('ds-flex ds-items-center ds-text-left', className)}>
      <input
        type="checkbox"
        id={label ? id : undefined}
        name={name}
        disabled={disabled}
        value={value}
        checked={checked}
        onChange={onChange}
        aria-label={label ? undefined : ariaLabel}
        className={cx(
          'ds-peer ds-relative ds-size-5 ds-min-h-5 ds-min-w-5 ds-appearance-none ds-rounded-none ds-bg-white',
          {
            ...borderVariantClassnames,
          },
        )}
      />
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="16"
        height="16"
        viewBox="0 0 16 16"
        fill="none"
        className="ds-pointer-events-none ds-absolute ds-hidden ds-fill-accent peer-checked:ds-block"
      >
        <rect x="0" y="0" width="16" height="16" />
      </svg>
      {label && (
        <label
          htmlFor={id}
          className={`ds-flex ds-flex-row ds-items-center ds-text-button-md ds-text-black peer-hover:ds-text-accent peer-hover:ds-underline peer-disabled:ds-text-inactive-gray peer-disabled:ds-no-underline ${!isLabelValidElement ? 'ds-pl-4' : ''}`.trim()}
        >
          {label}
        </label>
      )}
    </div>
  );
};
