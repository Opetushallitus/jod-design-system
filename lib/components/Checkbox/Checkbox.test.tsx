import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Checkbox } from './Checkbox';

const label = 'My Checkbox';
const myValue = 'myValue';

describe('Checkbox', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Checkbox name="myCheckbox" label={label} onChange={vi.fn} value={myValue} checked={false} />,
    );
    const checkbox = screen.getByLabelText(label);
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).not.toBeChecked();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles onChange event', () => {
    const handleChange = vi.fn();
    render(<Checkbox name="myCheckbox" label={label} onChange={handleChange} value={myValue} checked={false} />);
    const checkbox = screen.getByLabelText(label);
    fireEvent.click(checkbox);
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('disables the checkbox', () => {
    render(<Checkbox name="myCheckbox" label={label} onChange={vi.fn} value={myValue} checked={false} disabled />);
    const checkbox = screen.getByLabelText(label);
    expect(checkbox).toBeDisabled();
  });

  it('checks the checkbox', () => {
    render(<Checkbox name="myCheckbox" label={label} onChange={vi.fn} value={myValue} checked={true} />);
    const checkbox = screen.getByLabelText(label);
    expect(checkbox).toBeChecked();
  });
});
