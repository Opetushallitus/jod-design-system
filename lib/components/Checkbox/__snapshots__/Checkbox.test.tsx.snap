// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > renders correctly 1`] = `
<div
  class="ds:flex ds:items-center ds:text-left ds:relative"
>
  <input
    aria-checked="false"
    class="ds:peer ds:size-5 ds:min-h-5 ds:min-w-5 ds:appearance-none ds:rounded-none ds:bg-white ds:border-0"
    id="«r0»"
    name="myCheckbox"
    type="checkbox"
    value="myValue"
  />
  <svg
    class="ds:pointer-events-none ds:absolute ds:hidden ds:text-accent ds:peer-checked:block"
    data-state="visible"
    fill="none"
    height="16"
    style="display: none;"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4 8l3 3l5-5"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
    />
  </svg>
  <svg
    class="ds:pointer-events-none ds:absolute ds:fill-accent"
    fill="none"
    height="16"
    style="display: none;"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      height="2"
      width="10"
      x="3"
      y="7"
    />
  </svg>
  <label
    class="ds:flex ds:flex-row ds:items-center ds:text-heading-4 ds:text-black ds:peer-hover:text-accent ds:peer-hover:underline ds:peer-disabled:text-inactive-gray ds:peer-disabled:no-underline ds:pl-4"
    for="«r0»"
  >
    My Checkbox
  </label>
</div>
`;
